docs/DesignDocs/full_edl_serialization.md

```diff
@@ -0,0 +1,151 @@
+Enforcing full serialization in EDL
+===================================
+
+Motivation
+----------
+
+Currently EDL only provides partial support for parameter serialization. For
+example, the following EDL defines ECALLs whose parameters can only be
+partially serialized.
+
+```c++
+enclave
+{
+    /* Include struct iovec (foreign structure) */
+    #include <sys/uio.h>
+
+    trusted
+    {
+        public int send(
+            [in, count=iovcnt] const struct iovec* iov,
+            size_t iovcnt);
+
+        public int recv(
+            [out, count=iovcnt] struct iovec* iov,
+            size_t iovcnt);
+    };
+};
+```
+
+The generator produces code that serializes the array of **iovec**
+structures, but it cannot serialize the buffer referenced by the structure
+itself. Recall the definition of the **iovec** structure.
+
+```c++
+struct iovec
+{
+    void* iov_base;
+    size_t iov_len;
+};
+```
+
+Consequently, these ECALLs pass pointers to untrusted memory (**iov_base**) to
+the enclave developer's **send** and **recv** implementations.
+
+The exposure of untrusted pointers to the enclave implementation poses two main
+problems.
+
+- Implementations are vulnerable to coding errors that may introduce security
+  vulnerabilities.
+- The enclave application is not portable to trusted hardware environments
+  that do not support or have not enabled shared memory (e.g., TrustZone).
+
+Some potential security vulnerabilities include:
+
+- Time-of-check, Time-of-use errors (TOC-TOU).
+- Unwittingly writing secrets onto host memory.
+- Overwrite attacks where **iov_base** points to enclave memory.
+
+EDL provides a *deep-copy* feature for fully serializing parameters as shown
```

Comment:
@mikbras Could you clarify this for me? With deepcopy all structures should be completely serialized(Please correct me here if  I am wrong). Are there problems with the deepcopy implementation that would mean that structs would not be fully serialized? Please help me understand.

Reply:
This is the result of using `struct iovec` as-is (defined in the header, which does not have the deep-copy annotations on the struct). Contrast this to `struct local_iovec` which has the same internal structure, but with the necessary deep-copy annotations (`[size=iov_len]`).